File: worker/csj_worker/adapters/configs/mexc.py
"""MEXC — universal-adapter config.

API docs: https://mexcdevelop.github.io/apidocs/

Permission check
----------------
MEXC does not expose API key permissions on any read endpoint. The only
way to determine withdraw status is to *attempt* a withdraw — which is
destructive UX. Per the framework's documented policy (see
``adapters/configs/_base.py``) we treat MEXC as "withdraw-unverified" and
surface that to the UI so the user can confirm their key is read-only.

This is intentionally conservative: we never silently pass a destructive
key, but we cannot block one either. The UI MUST show the unverified
status and require user attestation.
"""

from __future__ import annotations

from typing import Any

from csj_worker.adapters.configs._base import VenueConfig


async def _fetch_permissions(client: Any) -> dict[str, Any]:
    """Probe Read access — withdraw scope is not introspectable."""
    balance = await client.fetch_balance()
    return {
        "verified_read": True,
        "withdraw_status": "unverified",
        "_raw_info": balance.get("info", {}) if isinstance(balance, dict) else {},
    }


def _has_withdraw(info: dict[str, Any]) -> bool:
    return False


def _extract_permissions(info: dict[str, Any]) -> list[str]:
    return ["read", "withdraw:unverified"]


CONFIG = VenueConfig(
    code="mexc",
    ccxt_id="mexc",
    ccxt_options={"options": {"defaultType": "swap"}},
    requires_passphrase=False,
    supports_spot=True,
    supports_perp=True,
    supports_dated_futures=False,
    supports_options=False,
    supports_funding_history=True,
    supports_open_positions=True,
    supports_klines=True,
    max_lookback_days=30,
    page_size=1000,
    market_types=("swap", "spot"),
    funding_market_types=("swap",),
    fetch_permissions=_fetch_permissions,
    has_withdraw_permission=_has_withdraw,
    extract_permissions=_extract_permissions,
    rate_limit_rps=5.0,
    rate_limit_burst=10,
    rate_limit_cooloff_seconds=30,
    api_docs_url="https://mexcdevelop.github.io/apidocs/",
    notes=(
        "MEXC has no permissions endpoint. The framework cannot verify "
        "withdraw status — the user must create a read-only key and the UI "
        "displays 'withdraw:unverified' to enforce attestation."
    ),
)

File: worker-ts/src/adapters/configs/mexc.ts
/**
 * MEXC — universal-adapter config (TS port of
 * `worker/csj_worker/adapters/configs/mexc.py`).
 *
 * Permission check
 * ----------------
 * MEXC exposes NO API-key-permission endpoint — withdraw scope cannot be
 * introspected without attempting a (destructive) withdraw. So, unlike
 * Bybit, `connect()` cannot reject a withdraw-capable MEXC key; it only
 * probes Read access via `fetchBalance` and surfaces `withdraw:unverified`
 * so the UI can require the user to attest the key is read-only.
 *
 * This is conservative-by-disclosure: the framework never silently passes
 * a destructive key, but it also cannot block one. The hard safety
 * guarantee is structural — `CcxtGenericAdapter` only ever calls read
 * methods (fetchMyTrades / fetchBalance / fetchFundingHistory /
 * fetchPositions / fetchOHLCV); it contains no trade or withdraw call.
 *
 * Market types
 * ------------
 * MEXC's ccxt class uses `defaultType` ('swap' | 'spot'). v1 iterates
 * `('swap', 'spot')`.
 */
import { defineVenueConfig, type VenueConfig } from '../venue-config.js';

/**
 * Probe Read access. A successful `fetchBalance` proves the key is valid
 * and can read — withdraw scope stays `unverified` (not introspectable).
 */
async function fetchPermissions(
  client: unknown,
): Promise<Record<string, unknown>> {
  const balance = (await (
    client as { fetchBalance: () => Promise<unknown> }
  ).fetchBalance()) as { info?: Record<string, unknown> } | undefined;
  const rawInfo =
    balance && typeof balance === 'object' && balance.info ? balance.info : {};
  return { verifiedRead: true, withdrawStatus: 'unverified', rawInfo };
}

/**
 * MEXC withdraw scope is not introspectable, so never auto-reject — the UI
 * enforces attestation via the `withdraw:unverified` permission string.
 */
function hasWithdraw(_info: Record<string, unknown>): boolean {
  return false;
}

function extractPermissions(_info: Record<string, unknown>): string[] {
  return ['read', 'withdraw:unverified'];
}

export const MEXC_CONFIG: VenueConfig = defineVenueConfig({
  code: 'mexc',
  ccxtId: 'mexc',
  ccxtOptions: { options: { defaultType: 'swap' } },
  requiresPassphrase: false,
  supportsSpot: true,
  supportsPerp: true,
  supportsDatedFutures: false,
  supportsOptions: false,
  supportsFundingHistory: true,
  supportsOpenPositions: true,
  supportsKlines: true,
  maxLookbackDays: 30,
  pageSize: 1000,
  marketTypes: ['swap', 'spot'],
  fundingMarketTypes: ['swap'],
  fetchPermissions,
  hasWithdrawPermission: hasWithdraw,
  extractPermissions,
  rateLimitRps: 5.0,
  rateLimitBurst: 10,
  rateLimitCooloffSeconds: 30,
  apiDocsUrl: 'https://mexcdevelop.github.io/apidocs/',
  notes:
    'MEXC has no permissions endpoint. The framework cannot verify ' +
    'withdraw status — the user must create a read-only key, and the UI ' +
    "displays 'withdraw:unverified' to enforce attestation.",
});
